### PR TITLE
Bug 1802

### LINKED ISSUE #1802: Allow accessing protected and private attributes in __new__
When implementing `__new__`, you usually create an instance of the class in which the `__new__` is defined. `pylint` currently does not allow setting protected and private attributes inside `__new__`.

### Steps to reproduce
Lint the follwing code

```python
# pylint: disable=blacklisted-name, missing-docstring

class MyObj:

    def __new__(cls, foo, bar):
        obj = object.__new__(cls)
        obj._foo = foo   # <-- warning reported here
        obj.__bar = bar   # <-- warning reported here
        return obj

    @property
    def foo(self):
        return self._foo

    @property
    def bar(self):
        return self.__bar



def main():
    instance = MyObj(1, 2)
    print(instance.foo)
    print(instance.bar)


main()
```

### Current behavior

This currently returns the following:

```
************* Module example
W:  7, 8: Access to a protected member _foo of a client class (protected-access)
W:  8, 8: Access to a protected member __bar of a client class (protected-access)

------------------------------------------------------------------
Your code has been rated at 8.67/10 (previous run: 7.69/10, +0.97)
```

### Expected behavior

No warning should be reported.

This may be controversial, but I wonder if it would make sense to silence the warnings in this case. just like they are inside of `__init__`. While `__new__` is used really rarely, it does not strike me as *wrong* to access private/protected members in there unless I am missing something. Would that really be a bad practice?

### pylint --version output

```
pylint 1.8.1,
astroid 1.6.0
Python 3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609]
```

This is also somewhat related to #1801 as I noticed this error with that enum, but it turns out that the warning is also emitted for normal objects.

### PR DESCRIPTION
**Pull request by hippo91**
## Steps

- [ * ] Add a ChangeLog entry describing what your PR does.
- [ * ] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [ * ] Write a good description on what the PR does.

## Description
This PR prevents the emission of **protected-access** message when a single underscore prefixed attribute is accessed inside a special method and after an `isinstance` test that ensures  the accessed object is of the same type that the current class.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |

## Related Issue

Closes #1802

### CODE REVIEW

**Comment by hippo91** [comment]
@PCManticore the **Travis** CI failed with `python3.8`.
The **AppVeyor** CI failed for `python3.4` whereas **Travis** CI is ok with it.

**Comment by hippo91** [comment]
@PCManticore ,
Apparently **AppVeyor** CI is broken. Can you confirm please?

**Comment by limdor** [comment]
Nice, to me looks reasonable

File: pylint/checkers/classes.py [hunk 1]
```diff
@@ -1308,8 +1309,74 @@ def _check_protected_attribute_access(self, node):
...
+        :note: test that at least a call to 'isinstance(node, self.__class__)' is made
+               in the parent If nodes
+        """
+        if_nodes_parent = ClassChecker._get_if_nodes_parent(node)
+        if not if_nodes_parent:
+            return False
+
+        # Check if one of the test made in the parent If nodes ensures that the node
+        #  is an instance of the current class.
+        for if_node in if_nodes_parent:
```
Comment (PCManticore) [inline pylint/checkers/classes.py]: @hippo91 I find this extra unnecessary complexity 😅 The implementation would be a lot simpler if we would just skip emitting `protected-access` inside a special method, assuming the person that wrote the code already handled the case of accessing the attribute on a same instance.

**Comment by hippo91** [comment]
@PCManticore thanks for your remarks. I took them into account. I had to modify the `test_uninferable_attribute` method of the `unittest_checker_classes` module.
Are you ok with it?

**Comment by PCManticore** [comment]
@hippo91 Great stuff!

**Pull request merged**